In [1]:
import pandas as pd
import numpy as np

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from gensim.models import Word2Vec
from gensim.models import FastText
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity

from huggingface_hub import hf_hub_download
from gensim.models import KeyedVectors
import fasttext
import nltk
nltk.download("punkt")
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\PC\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\PC\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

In [2]:
courses_df = pd.read_csv('take-home-2/courses.csv')

courses_df = courses_df.fillna('')
courses_df['text'] = courses_df[['preprocessed_title', 'preprocessed_description', 'preprocessed_department_name', 'preprocessed_topics']].values.tolist()

# topics and department names are separated by comma, we don't want that in the text
courses_df['text'] = [' '.join(t).replace(', ', '') for t in courses_df['text']]

# Build Vectors & Models

In [ ]:
# BoW
def bow_embeddings(corpus):
    count_vectorizer = CountVectorizer()
    embeddings_bow = count_vectorizer.fit_transform(corpus)
    return embeddings_bow, count_vectorizer

# TF-IDF
def tfidf_embeddings(corpus):
    tfidf_vectorizer = TfidfVectorizer()
    embeddings_tfidf = tfidf_vectorizer.fit_transform(corpus)
    return embeddings_tfidf, tfidf_vectorizer

# Word2Vec
def word2vec_model(corpus):
    kv = KeyedVectors.load_word2vec_format("take-home-2/wiki-news-300d-1M.vec")
    
    # only retrieving necessary tokens, to save memory
    all_tokens = set()
    
    for text in corpus:
        tokens = nltk.word_tokenize(text)
        all_tokens.update(tokens)
        
    filtered_tokens = [token for token in all_tokens if token in kv]
    
    kv_subset = KeyedVectors(vector_size=kv.vector_size)
    kv_subset.add_vectors(filtered_tokens, [kv[token] for token in filtered_tokens])
    return kv_subset
    
# FastText
def fasttext_model(corpus):
    model = FastText(corpus, vector_size=100, window=5, min_count=1, workers=-1)
    
    return model.wv

    # Wanted to use this, but not enough RAM
    
    # model_path = hf_hub_download(repo_id="facebook/fasttext-en-vectors", filename="model.bin")
    # ft_model = fasttext.load_model(model_path)
    
    # vector_size = ft_model.get_dimension()
    # kv = KeyedVectors(vector_size=vector_size)
    
    # all_tokens = set()
    
    # for text in corpus:
    #     tokens = nltk.word_tokenize(text)
    #     all_tokens.update(tokens)

    # word_vectors = []
    # words_in_model = []
    # for token in all_tokens:
    #     try:
    #         word_vectors.append(ft_model.get_word_vector(token))
    #         words_in_model.append(token)
    #     except:
    #         pass

    # kv.add_vectors(words_in_model, word_vectors)
    # return kv

# Sentence Transformers
def sentence_transformers_embeddings(corpus):
    model = SentenceTransformer("all-MiniLM-L6-v2")
    return model.encode(corpus), model

In [22]:
# search courses similar to this given course

courses_df.loc[1797, ['title', 'description', 'department_name', 'topics']].to_dict()

{'title': 'Machine Learning',
 'description': '6.867 is an introductory course on machine learning which gives an overview of many concepts, techniques, and algorithms in machine learning, beginning with topics such as classification and linear regression and ending up with more recent topics such as boosting, support vector machines, hidden Markov models, and Bayesian networks. The course will give the student the basic ideas and intuition behind modern machine learning methods as well as a bit more formal understanding of how, why, and when they work. The underlying theme in the course is statistical inference as it provides the foundation for most of the methods covered.',
 'department_name': 'Electrical Engineering and Computer Science',
 'topics': 'Engineering, Science, Computer Science, Artificial Intelligence, Mathematics, Probability and Statistics, Cognitive Science, Algorithms and Data Structures'}

In [24]:
courses_df.loc[1797, ['text']].to_dict()

{'text': 'machine learning 6867 introductory course machine learning give overview many concept technique algorithm machine learning beginning topic classification linear regression ending recent topic boosting support vector machine hidden markov model bayesian network course give student basic idea intuition behind modern machine learning method well bit formal understanding work underlying theme course statistical inference provides foundation method covered electrical engineering computer science engineeringsciencecomputer scienceartificial intelligencemathematicsprobability statisticcognitive sciencealgorithm data structure'}

# BoW x Cosine

In [ ]:
sentence = courses_df.loc[1797, 'text']

corpus_embeddings, vectorizer = bow_embeddings(courses_df['text'])
pd.DataFrame({"text": courses_df['text'], "bow_vectors": corpus_embeddings.toarray().tolist()}).to_csv('take-home-2/vectors/bow_vectors.csv', index=False)
# to reverse the cos similarity value, so that 0 is the most similar
dist = 1-cosine_similarity(vectorizer.transform([sentence]), corpus_embeddings).flatten()
bow_cos_dist = courses_df.iloc[np.argsort(dist), :]
bow_cos_dist.loc[:,['cos_dist']] = np.sort(dist)
bow_cos_dist.to_csv('take-home-2/similarities/bow_cos_dist.csv', index=False)
bow_cos_dist[:11]

number                                              title  \
1797    6.867                                   Machine Learning   
1218   18.465  Topics in Statistics: Statistical Learning Theory   
715     6.036                   Introduction to Machine Learning   
1756   18.409            Algorithmic Aspects of Machine Learning   
1519   6.S897                    Machine Learning for Healthcare   
1697   18.065  Matrix Methods in Data Analysis, Signal Proces...   
173    15.097        Prediction: Machine Learning and Statistics   
2134  9.520-A  Networks for Learning: Regression and Classifi...   
259    18.657                    Mathematics of Machine Learning   
154     9.520       Statistical Learning Theory and Applications   
1780    6.825   Techniques in Artificial Intelligence (SMA 5504)   

              level semester    year  \
1797       Graduate     Fall  2006.0   
1218       Graduate   Spring  2007.0   
715   Undergraduate     Fall  2020.0   
1756       Graduate   Spring  2015.0   
1519       Graduate   Spring  2019.0   
1697  Undergraduate   Spring  2018.0   
173        Graduate   Spring  2012.0   
2134       Graduate   Spring  2001.0   
259        Graduate     Fall  2015.0   
154        Graduate   Spring  2006.0   
1780       Graduate     Fall  2002.0   

                                            description  \
1797  6.867 is an introductory course on machine lea...   
1218  The main goal of this course is to study the g...   
715   This course introduces principles, algorithms,...   
1756  This course is organized around algorithmic is...   
1519  This course introduces students to machine lea...   
1697  Linear algebra concepts are key for understand...   
173   Prediction is at the heart of almost every sci...   
2134  The course focuses on the problem of supervise...   
259   Broadly speaking, Machine Learning refers to t...   
154   This course is for upper-level graduate studen...   
1780  6.825 is a graduate-level introduction to arti...   

                                                 topics  \
1797  Engineering, Science, Computer Science, Artifi...   
1218  Engineering, Science, Computer Science, Artifi...   
715   Engineering, Computer Science, Artificial Inte...   
1756  Engineering, Computer Science, Mathematics, Ap...   
1519  Engineering, Computer Science, Artificial Inte...   
1697  Engineering, Mathematics, Electrical Engineeri...   
173   Engineering, Computer Science, Artificial Inte...   
2134  Science, Mathematics, Probability and Statisti...   
259   Engineering, Computer Science, Artificial Inte...   
154   Science, Mathematics, Probability and Statisti...   
1780  Engineering, Science, Computer Science, Artifi...   

                                            instructors  \
1797    Prof. Tommi Jaakkola, Ali Mohammad, Rohit Singh   
1218                             Prof. Dmitry Panchenko   
715   Prof. Duane Boning, Prof. Isaac Chuang, Prof. ...   
1756                                 Prof. Ankur Moitra   
1519          Prof. David Sontag, Prof. Peter Szolovits   
1697                               Prof. Gilbert Strang   
173                                 Prof. Cynthia Rudin   
2134          Prof. Tomaso Poggio, Dr. Alessandro Verri   
259                             Prof. Philippe Rigollet   
154                                 Prof. Tomaso Poggio   
1780   Prof. Leslie Kaelbling, Prof. Tomás Lozano-Pérez   

                                        department_name  \
1797        Electrical Engineering and Computer Science   
1218                                        Mathematics   
715         Electrical Engineering and Computer Science   
1756                                        Mathematics   
1519  Electrical Engineering and Computer Science, H...   
1697                                        Mathematics   
173                          Sloan School of Management   
2134                       Brain and Cognitive Sciences   
259                                         Mathema

# TF-IDF x Cosine

In [ ]:
sentence = courses_df.loc[1797, 'text']

corpus_embeddings, vectorizer = tfidf_embeddings(courses_df['text'])
pd.DataFrame({"text": courses_df['text'], "tfidf_vectors": corpus_embeddings.toarray().tolist()}).to_csv('take-home-2/vectors/tfidf_vectors.csv', index=False)
# to reverse the cos similarity value, so that 0 is the most similar
dist = 1-cosine_similarity(vectorizer.transform([sentence]), corpus_embeddings).flatten()
tfidf_cos_dist = courses_df.iloc[np.argsort(dist), :]
tfidf_cos_dist.loc[:,['cos_dist']] = np.sort(dist)
tfidf_cos_dist.to_csv('take-home-2/similarities/tfidf_cos_dist.csv', index=False)
tfidf_cos_dist[:11]

number                                              title  \
1797    6.867                                   Machine Learning   
1218   18.465  Topics in Statistics: Statistical Learning Theory   
1756   18.409            Algorithmic Aspects of Machine Learning   
607     9.520       Statistical Learning Theory and Applications   
715     6.036                   Introduction to Machine Learning   
173    15.097        Prediction: Machine Learning and Statistics   
2134  9.520-A  Networks for Learning: Regression and Classifi...   
1697   18.065  Matrix Methods in Data Analysis, Signal Proces...   
1519   6.S897                    Machine Learning for Healthcare   
154     9.520       Statistical Learning Theory and Applications   
259    18.657                    Mathematics of Machine Learning   

              level semester    year  \
1797       Graduate     Fall  2006.0   
1218       Graduate   Spring  2007.0   
1756       Graduate   Spring  2015.0   
607        Graduate   Spring  2003.0   
715   Undergraduate     Fall  2020.0   
173        Graduate   Spring  2012.0   
2134       Graduate   Spring  2001.0   
1697  Undergraduate   Spring  2018.0   
1519       Graduate   Spring  2019.0   
154        Graduate   Spring  2006.0   
259        Graduate     Fall  2015.0   

                                            description  \
1797  6.867 is an introductory course on machine lea...   
1218  The main goal of this course is to study the g...   
1756  This course is organized around algorithmic is...   
607   Focuses on the problem of supervised learning ...   
715   This course introduces principles, algorithms,...   
173   Prediction is at the heart of almost every sci...   
2134  The course focuses on the problem of supervise...   
1697  Linear algebra concepts are key for understand...   
1519  This course introduces students to machine lea...   
154   This course is for upper-level graduate studen...   
259   Broadly speaking, Machine Learning refers to t...   

                                                 topics  \
1797  Engineering, Science, Computer Science, Artifi...   
1218  Engineering, Science, Computer Science, Artifi...   
1756  Engineering, Computer Science, Mathematics, Ap...   
607   Engineering, Science, Computer Science, Artifi...   
715   Engineering, Computer Science, Artificial Inte...   
173   Engineering, Computer Science, Artificial Inte...   
2134  Science, Mathematics, Probability and Statisti...   
1697  Engineering, Mathematics, Electrical Engineeri...   
1519  Engineering, Computer Science, Artificial Inte...   
154   Science, Mathematics, Probability and Statisti...   
259   Engineering, Computer Science, Artificial Inte...   

                                            instructors  \
1797    Prof. Tommi Jaakkola, Ali Mohammad, Rohit Singh   
1218                             Prof. Dmitry Panchenko   
1756                                 Prof. Ankur Moitra   
607   Dr. Sayan Mukherjee, Prof. Tomaso Poggio, Alex...   
715   Prof. Duane Boning, Prof. Isaac Chuang, Prof. ...   
173                                 Prof. Cynthia Rudin   
2134          Prof. Tomaso Poggio, Dr. Alessandro Verri   
1697                               Prof. Gilbert Strang   
1519          Prof. David Sontag, Prof. Peter Szolovits   
154                                 Prof. Tomaso Poggio   
259                             Prof. Philippe Rigollet   

                                        department_name  \
1797        Electrical Engineering and Computer Science   
1218                                        Mathematics   
1756                                        Mathematics   
607                        Brain and Cognitive Sciences   
715         Electrical Engineering and Computer Science   
173                          Sloan School of Management   
2134                       Brain and Cognitive Sciences   
1697                                        Mathematics   
1519  Electrical Engineering and Computer Science, 

# Word2Vec x WMD

In [ ]:
sentence = courses_df.loc[1797, 'text']

kv = word2vec_model(courses_df['text'])
pd.DataFrame({"word": kv.index_to_key, "vectors": [kv[word] for word in kv.index_to_key]}).to_csv('take-home-2/vectors/w2v_word_vectors.csv', index=False)
dists = []
for text in courses_df['text']:
    # distance between two list of word
    dist = kv.wmdistance(nltk.word_tokenize(sentence), nltk.word_tokenize(text))
    dists.append(dist)

w2v_wmd_dist = courses_df.iloc[np.argsort(dists), :]
w2v_wmd_dist.loc[:,['wmd_dist']] = np.sort(dists)
w2v_wmd_dist.to_csv('take-home-2/similarities/w2v_wmd_dist.csv', index=False)
w2v_wmd_dist[:11]

number                                              title  \
1797     6.867                                   Machine Learning   
1780     6.825   Techniques in Artificial Intelligence (SMA 5504)   
1218    18.465  Topics in Statistics: Statistical Learning Theory   
154      9.520       Statistical Learning Theory and Applications   
2134   9.520-A  Networks for Learning: Regression and Classifi...   
715      6.036                   Introduction to Machine Learning   
1697    18.065  Matrix Methods in Data Analysis, Signal Proces...   
607      9.520       Statistical Learning Theory and Applications   
2286     6.438                           Algorithms for Inference   
2136  HST.951J                           Medical Decision Support   
1519    6.S897                    Machine Learning for Healthcare   

              level semester    year  \
1797       Graduate     Fall  2006.0   
1780       Graduate     Fall  2002.0   
1218       Graduate   Spring  2007.0   
154        Graduate   Spring  2006.0   
2134       Graduate   Spring  2001.0   
715   Undergraduate     Fall  2020.0   
1697  Undergraduate   Spring  2018.0   
607        Graduate   Spring  2003.0   
2286       Graduate     Fall  2014.0   
2136       Graduate     Fall  2005.0   
1519       Graduate   Spring  2019.0   

                                            description  \
1797  6.867 is an introductory course on machine lea...   
1780  6.825 is a graduate-level introduction to arti...   
1218  The main goal of this course is to study the g...   
154   This course is for upper-level graduate studen...   
2134  The course focuses on the problem of supervise...   
715   This course introduces principles, algorithms,...   
1697  Linear algebra concepts are key for understand...   
607   Focuses on the problem of supervised learning ...   
2286  This is a graduate-level introduction to the p...   
2136  This course presents the main concepts of deci...   
1519  This course introduces students to machine lea...   

                                                 topics  \
1797  Engineering, Science, Computer Science, Artifi...   
1780  Engineering, Science, Computer Science, Artifi...   
1218  Engineering, Science, Computer Science, Artifi...   
154   Science, Mathematics, Probability and Statisti...   
2134  Science, Mathematics, Probability and Statisti...   
715   Engineering, Computer Science, Artificial Inte...   
1697  Engineering, Mathematics, Electrical Engineeri...   
607   Engineering, Science, Computer Science, Artifi...   
2286  Engineering, Computer Science, Mathematics, Pr...   
2136  Engineering, Computer Science, Artificial Inte...   
1519  Engineering, Computer Science, Artificial Inte...   

                                            instructors  \
1797    Prof. Tommi Jaakkola, Ali Mohammad, Rohit Singh   
1780   Prof. Leslie Kaelbling, Prof. Tomás Lozano-Pérez   
1218                             Prof. Dmitry Panchenko   
154                                 Prof. Tomaso Poggio   
2134          Prof. Tomaso Poggio, Dr. Alessandro Verri   
715   Prof. Duane Boning, Prof. Isaac Chuang, Prof. ...   
1697                               Prof. Gilbert Strang   
607   Dr. Sayan Mukherjee, Prof. Tomaso Poggio, Alex...   
2286                                Prof. Devavrat Shah   
2136    Prof. Lucila Ohno-Machado, Prof. Staal Vinterbo   
1519          Prof. David Sontag, Prof. Peter Szolovits   

                                        department_name  \
1797        Electrical Engineering and Computer Science   
1780        Electrical Engineering and Computer Science   
1218                                        Mathematics   
154                        Brain and Cognitive Sciences   
2134                       Brain and Cognitive Sciences   
715         Electrical Engineering and Computer Science   
1697                                        Mathematics   
607                        Brain and Cognitive Sciences   
2286        Electrical Engineering and C

# Word2Vec x Cosine

Instead of calculating distance by word, we can try to calculate distance by sentence. We need to convert a list of vectors for each document into a single vector by averaging the vectors value.

In [ ]:
def avg_vector(vectors:list, n_words:int) -> list:
    # Calculate the average vector from a list of vectors
    vector_sum = np.sum(vectors, axis=0)
    if n_words != 0:
        return np.divide(vector_sum, n_words)

In [ ]:
corpus_vectors = []

# Create a list of vectors for each text in the corpus
for text in courses_df['text']:
    tokens = nltk.word_tokenize(text)
    vectors = []
    for token in tokens:
        if token in kv:
            vectors.append(kv[token])
        else:
            # if the token is not in the model, we use a vector of ones
            vectors.append(np.ones(kv.vector_size, dtype='float32'))
    corpus_vectors.append(vectors)

# Average the word vectors for each text
corpus_vectors = [avg_vector(vector, len(vector)) for vector in corpus_vectors]

pd.DataFrame({"text": courses_df['text'], "w2v_sentence_vector": corpus_vectors}).to_csv('take-home-2/vectors/w2v_sentence_vectors.csv', index=False)

sentence = courses_df.loc[1797, 'text']
sentence_tokens = nltk.word_tokenize(sentence)
sentence_vector = [avg_vector([kv[token] if token in kv else np.ones(kv.vector_size, dtype='float32') for token in sentence_tokens], len(sentence_tokens))]

# reverse the cos similarity value, so that 0 is the most similar
dist = 1-cosine_similarity(sentence_vector, corpus_vectors).flatten()
w2v_cos_dist = courses_df.iloc[np.argsort(dist), :]
w2v_cos_dist.loc[:,['cos_dist']] = np.sort(dist)
w2v_cos_dist.to_csv('take-home-2/similarities/w2v_cos_dist.csv', index=False)
w2v_cos_dist[:11]

number                                              title  \
1797      6.867                                   Machine Learning   
1631     6.S191                      Introduction to Deep Learning   
607       9.520       Statistical Learning Theory and Applications   
2001   MAS.622J                   Pattern Recognition and Analysis   
715       6.036                   Introduction to Machine Learning   
98        6.041  Probabilistic Systems Analysis and Applied Pro...   
2245      6.080        Great Ideas in Theoretical Computer Science   
1705     6.01SC  Introduction to Electrical Engineering and Com...   
721      6.S980              Machine Learning for Inverse Graphics   
1699  RES.9-009  Introduction to Computational Neuroscience wit...   
2540     6.S062  Generative Artificial Intelligence in K–12 Edu...   

                     level     semester    year  \
1797              Graduate         Fall  2006.0   
1631         Undergraduate  January IAP  2020.0   
607               Graduate       Spring  2003.0   
2001              Graduate         Fall  2006.0   
715          Undergraduate         Fall  2020.0   
98           Undergraduate         Fall  2010.0   
2245         Undergraduate       Spring  2008.0   
1705         Undergraduate       Spring  2011.0   
721               Graduate         Fall  2022.0   
1699  Graduate, Non-Credit  January IAP  2025.0   
2540         Undergraduate         Fall  2023.0   

                                            description  \
1797  6.867 is an introductory course on machine lea...   
1631  This is MIT's introductory course on deep lear...   
607   Focuses on the problem of supervised learning ...   
2001  This class deals with the fundamentals of char...   
715   This course introduces principles, algorithms,...   
98    Welcome to 6.041/6.431, a subject on the model...   
2245  This course provides a challenging introductio...   
1705  This course provides an integrated introductio...   
721   This course covers fundamental and advanced te...   
1699  In this course, you will learn the basics of c...   
2540  The emergence of transformer architectures in ...   

                                                 topics  \
1797  Engineering, Science, Computer Science, Artifi...   
1631  Engineering, Computer Science, Artificial Inte...   
607   Engineering, Science, Computer Science, Artifi...   
2001  Engineering, Science, Mathematics, Probability...   
715   Engineering, Computer Science, Artificial Inte...   
98    Engineering, Mathematics, Systems Engineering,...   
2245  Engineering, Computer Science, Mathematics, Ap...   
1705  Engineering, Computer Science, Artificial Inte...   
721   Engineering, Computer Science, Artificial Inte...   
1699  Engineering, Science, Computer Science, Health...   
2540  Engineering, Computer Science, Artificial Inte...   

                                            instructors  \
1797    Prof. Tommi Jaakkola, Ali Mohammad, Rohit Singh   
1631                     Alexander Amini, Ava Soleimany   
607   Dr. Sayan Mukherjee, Prof. Tomaso Poggio, Alex...   
2001  Media Lab Faculty and Staff, Bo Morgan, Prof. ...   
715   Prof. Duane Boning, Prof. Isaac Chuang, Prof. ...   
98                                Prof. John Tsitsiklis   
2245                               Prof. Scott Aaronson   
1705  Prof. Harold Abelson, Prof. Isaac Chuang, Prof...   
721                              Prof. Vincent Sitzmann   
1699  Prof. Lilianne Mujica-Parodi, Dr. Haris Organt...   
2540  Prof. Harold Abelson, Safinah Ali, Prof. Cynth...   

                                        department_name  \
1797        Electrical Engineering and Computer Science   
1631        Electrical Engineering and Computer Science   
607                        Brain and Cognitive Sciences   
2001  Media Arts and Sciences, Civil and Environment...   
715         Electrical Engineering and Computer Science   
98          Electrical Engineering and Computer Science   
2245        Ele

# FastText x WMD

In [14]:
sentence = courses_df.loc[1797, 'text']

ft = fasttext_model(courses_df['text'])
pd.DataFrame({"word": ft.index_to_key, "vectors": [ft[word] for word in ft.index_to_key]}).to_csv('take-home-2/vectors/ft_word_vectors.csv', index=False)
dists = []
for text in courses_df['text']:
    dist = ft.wmdistance(nltk.word_tokenize(sentence), nltk.word_tokenize(text))
    dists.append(dist)
    
ft_wmd_dist = courses_df.iloc[np.argsort(dists), :]
ft_wmd_dist.loc[:,['wmd_dist']] = np.sort(dists)
ft_wmd_dist.to_csv('take-home-2/similarities/ft_wmd_dist.csv', index=False)
ft_wmd_dist[:11]

number                                              title  \
1797    6.867                                   Machine Learning   
1780    6.825   Techniques in Artificial Intelligence (SMA 5504)   
1218   18.465  Topics in Statistics: Statistical Learning Theory   
607     9.520       Statistical Learning Theory and Applications   
715     6.036                   Introduction to Machine Learning   
1519   6.S897                    Machine Learning for Healthcare   
154     9.520       Statistical Learning Theory and Applications   
259    18.657                    Mathematics of Machine Learning   
2134  9.520-A  Networks for Learning: Regression and Classifi...   
1697   18.065  Matrix Methods in Data Analysis, Signal Proces...   
173    15.097        Prediction: Machine Learning and Statistics   

              level semester    year  \
1797       Graduate     Fall  2006.0   
1780       Graduate     Fall  2002.0   
1218       Graduate   Spring  2007.0   
607        Graduate   Spring  2003.0   
715   Undergraduate     Fall  2020.0   
1519       Graduate   Spring  2019.0   
154        Graduate   Spring  2006.0   
259        Graduate     Fall  2015.0   
2134       Graduate   Spring  2001.0   
1697  Undergraduate   Spring  2018.0   
173        Graduate   Spring  2012.0   

                                            description  \
1797  6.867 is an introductory course on machine lea...   
1780  6.825 is a graduate-level introduction to arti...   
1218  The main goal of this course is to study the g...   
607   Focuses on the problem of supervised learning ...   
715   This course introduces principles, algorithms,...   
1519  This course introduces students to machine lea...   
154   This course is for upper-level graduate studen...   
259   Broadly speaking, Machine Learning refers to t...   
2134  The course focuses on the problem of supervise...   
1697  Linear algebra concepts are key for understand...   
173   Prediction is at the heart of almost every sci...   

                                                 topics  \
1797  Engineering, Science, Computer Science, Artifi...   
1780  Engineering, Science, Computer Science, Artifi...   
1218  Engineering, Science, Computer Science, Artifi...   
607   Engineering, Science, Computer Science, Artifi...   
715   Engineering, Computer Science, Artificial Inte...   
1519  Engineering, Computer Science, Artificial Inte...   
154   Science, Mathematics, Probability and Statisti...   
259   Engineering, Computer Science, Artificial Inte...   
2134  Science, Mathematics, Probability and Statisti...   
1697  Engineering, Mathematics, Electrical Engineeri...   
173   Engineering, Computer Science, Artificial Inte...   

                                            instructors  \
1797    Prof. Tommi Jaakkola, Ali Mohammad, Rohit Singh   
1780   Prof. Leslie Kaelbling, Prof. Tomás Lozano-Pérez   
1218                             Prof. Dmitry Panchenko   
607   Dr. Sayan Mukherjee, Prof. Tomaso Poggio, Alex...   
715   Prof. Duane Boning, Prof. Isaac Chuang, Prof. ...   
1519          Prof. David Sontag, Prof. Peter Szolovits   
154                                 Prof. Tomaso Poggio   
259                             Prof. Philippe Rigollet   
2134          Prof. Tomaso Poggio, Dr. Alessandro Verri   
1697                               Prof. Gilbert Strang   
173                                 Prof. Cynthia Rudin   

                                        department_name  \
1797        Electrical Engineering and Computer Science   
1780        Electrical Engineering and Computer Science   
1218                                        Mathematics   
607                        Brain and Cognitive Sciences   
715         Electrical Engineering and Computer Science   
1519  Electrical Engineering and Computer Science, H...   
154                        Brain and Cognitive Sciences   
259                                         Mathematics   
2134                       Brain and Cognitive Scie

# Sentence Transformers x Cosine

In [ ]:
sentence = courses_df.loc[1797, 'text']

corpus_embeddings, model = sentence_transformers_embeddings(courses_df['text'])
pd.DataFrame({"text": courses_df['text'], "st_vectors": corpus_embeddings.tolist()}).to_csv('take-home-2/vectors/st_vectors.csv', index=False)
# reverse the cos similarity value, so that 0 is the most similar
dists = 1-cosine_similarity([model.encode(sentence)], corpus_embeddings).flatten()
st_cos_dist = courses_df.iloc[np.argsort(dists), :]
st_cos_dist.loc[:,['cos_dist']] = np.sort(dists)
st_cos_dist.to_csv('take-home-2/similarities/st_cos_dist.csv', index=False)
st_cos_dist[:11]

number                                              title  \
1797     6.867                                   Machine Learning   
1218    18.465  Topics in Statistics: Statistical Learning Theory   
1780     6.825   Techniques in Artificial Intelligence (SMA 5504)   
2001  MAS.622J                   Pattern Recognition and Analysis   
715      6.036                   Introduction to Machine Learning   
259     18.657                    Mathematics of Machine Learning   
2134   9.520-A  Networks for Learning: Regression and Classifi...   
173     15.097        Prediction: Machine Learning and Statistics   
154      9.520       Statistical Learning Theory and Applications   
1697    18.065  Matrix Methods in Data Analysis, Signal Proces...   
1564     6.034                            Artificial Intelligence   

              level semester    year  \
1797       Graduate     Fall  2006.0   
1218       Graduate   Spring  2007.0   
1780       Graduate     Fall  2002.0   
2001       Graduate     Fall  2006.0   
715   Undergraduate     Fall  2020.0   
259        Graduate     Fall  2015.0   
2134       Graduate   Spring  2001.0   
173        Graduate   Spring  2012.0   
154        Graduate   Spring  2006.0   
1697  Undergraduate   Spring  2018.0   
1564  Undergraduate   Spring  2005.0   

                                            description  \
1797  6.867 is an introductory course on machine lea...   
1218  The main goal of this course is to study the g...   
1780  6.825 is a graduate-level introduction to arti...   
2001  This class deals with the fundamentals of char...   
715   This course introduces principles, algorithms,...   
259   Broadly speaking, Machine Learning refers to t...   
2134  The course focuses on the problem of supervise...   
173   Prediction is at the heart of almost every sci...   
154   This course is for upper-level graduate studen...   
1697  Linear algebra concepts are key for understand...   
1564  This course introduces representations, techni...   

                                                 topics  \
1797  Engineering, Science, Computer Science, Artifi...   
1218  Engineering, Science, Computer Science, Artifi...   
1780  Engineering, Science, Computer Science, Artifi...   
2001  Engineering, Science, Mathematics, Probability...   
715   Engineering, Computer Science, Artificial Inte...   
259   Engineering, Computer Science, Artificial Inte...   
2134  Science, Mathematics, Probability and Statisti...   
173   Engineering, Computer Science, Artificial Inte...   
154   Science, Mathematics, Probability and Statisti...   
1697  Engineering, Mathematics, Electrical Engineeri...   
1564  Engineering, Computer Science, Artificial Inte...   

                                            instructors  \
1797    Prof. Tommi Jaakkola, Ali Mohammad, Rohit Singh   
1218                             Prof. Dmitry Panchenko   
1780   Prof. Leslie Kaelbling, Prof. Tomás Lozano-Pérez   
2001  Media Lab Faculty and Staff, Bo Morgan, Prof. ...   
715   Prof. Duane Boning, Prof. Isaac Chuang, Prof. ...   
259                             Prof. Philippe Rigollet   
2134          Prof. Tomaso Poggio, Dr. Alessandro Verri   
173                                 Prof. Cynthia Rudin   
154                                 Prof. Tomaso Poggio   
1697                               Prof. Gilbert Strang   
1564   Prof. Leslie Kaelbling, Prof. Tomás Lozano-Pérez   

                                        department_name  \
1797        Electrical Engineering and Computer Science   
1218                                        Mathematics   
1780        Electrical Engineering and Computer Science   
2001  Media Arts and Sciences, Civil and Environment...   
715         Electrical Engineering and Computer Science   
259                                         Mathematics   
2134                       Brain and Cognitive Sciences   
173                          Sloan School of Management   
154                        Brain and Cog

# Model Evaluation

We don't have the ground truth, so we will manually check relevancy of recommended items in all methods

## Traditional Methods (BoW, TF-IDF)

In [ ]:
pd.set_option('display.max_rows', 11)

print("BoW x Cosine Method:")
display(bow_cos_dist.loc[:11, ['title', 'description', 'topics', 'department_name', 'cos_dist']])

print("TFIDF x Cosine Method:")
display(tfidf_cos_dist.loc[:11, ['title', 'description', 'topics', 'department_name', 'cos_dist']])

BoW x Cosine Method:


title  \
1797                                   Machine Learning   
1218  Topics in Statistics: Statistical Learning Theory   
715                    Introduction to Machine Learning   
1756            Algorithmic Aspects of Machine Learning   
1519                    Machine Learning for Healthcare   
1697  Matrix Methods in Data Analysis, Signal Proces...   
173         Prediction: Machine Learning and Statistics   
2134  Networks for Learning: Regression and Classifi...   
259                     Mathematics of Machine Learning   
154        Statistical Learning Theory and Applications   
1780   Techniques in Artificial Intelligence (SMA 5504)   

                                            description  \
1797  6.867 is an introductory course on machine lea...   
1218  The main goal of this course is to study the g...   
715   This course introduces principles, algorithms,...   
1756  This course is organized around algorithmic is...   
1519  This course introduces students to machine lea...   
1697  Linear algebra concepts are key for understand...   
173   Prediction is at the heart of almost every sci...   
2134  The course focuses on the problem of supervise...   
259   Broadly speaking, Machine Learning refers to t...   
154   This course is for upper-level graduate studen...   
1780  6.825 is a graduate-level introduction to arti...   

                                                 topics  \
1797  Engineering, Science, Computer Science, Artifi...   
1218  Engineering, Science, Computer Science, Artifi...   
715   Engineering, Computer Science, Artificial Inte...   
1756  Engineering, Computer Science, Mathematics, Ap...   
1519  Engineering, Computer Science, Artificial Inte...   
1697  Engineering, Mathematics, Electrical Engineeri...   
173   Engineering, Computer Science, Artificial Inte...   
2134  Science, Mathematics, Probability and Statisti...   
259   Engineering, Computer Science, Artificial Inte...   
154   Science, Mathematics, Probability and Statisti...   
1780  Engineering, Science, Computer Science, Artifi...   

                                        department_name      cos_dist  
1797        Electrical Engineering and Computer Science  6.661338e-16  
1218                                        Mathematics  5.113727e-01  
715         Electrical Engineering and Computer Science  5.135772e-01  
1756                                        Mathematics  5.199206e-01  
1519  Electrical Engineering and Computer Science, H...  5.409302e-01  
1697                                        Mathematics  5.546331e-01  
173                          Sloan School of Management  5.607724e-01  
2134                       Brain and Cognitive Sciences  5.937594e-01  
259                                         Mathematics  6.096400e-01  
154                        Brain and Cognitive Sciences  6.235877e-01  
1780        Electrical Engineering and Computer Science  6.251711e-01

TFIDF x Cosine Method:


title  \
1797                                   Machine Learning   
1218  Topics in Statistics: Statistical Learning Theory   
1756            Algorithmic Aspects of Machine Learning   
607        Statistical Learning Theory and Applications   
715                    Introduction to Machine Learning   
173         Prediction: Machine Learning and Statistics   
2134  Networks for Learning: Regression and Classifi...   
1697  Matrix Methods in Data Analysis, Signal Proces...   
1519                    Machine Learning for Healthcare   
154        Statistical Learning Theory and Applications   
259                     Mathematics of Machine Learning   

                                            description  \
1797  6.867 is an introductory course on machine lea...   
1218  The main goal of this course is to study the g...   
1756  This course is organized around algorithmic is...   
607   Focuses on the problem of supervised learning ...   
715   This course introduces principles, algorithms,...   
173   Prediction is at the heart of almost every sci...   
2134  The course focuses on the problem of supervise...   
1697  Linear algebra concepts are key for understand...   
1519  This course introduces students to machine lea...   
154   This course is for upper-level graduate studen...   
259   Broadly speaking, Machine Learning refers to t...   

                                                 topics  \
1797  Engineering, Science, Computer Science, Artifi...   
1218  Engineering, Science, Computer Science, Artifi...   
1756  Engineering, Computer Science, Mathematics, Ap...   
607   Engineering, Science, Computer Science, Artifi...   
715   Engineering, Computer Science, Artificial Inte...   
173   Engineering, Computer Science, Artificial Inte...   
2134  Science, Mathematics, Probability and Statisti...   
1697  Engineering, Mathematics, Electrical Engineeri...   
1519  Engineering, Computer Science, Artificial Inte...   
154   Science, Mathematics, Probability and Statisti...   
259   Engineering, Computer Science, Artificial Inte...   

                                        department_name  cos_dist  
1797        Electrical Engineering and Computer Science  0.000000  
1218                                        Mathematics  0.525189  
1756                                        Mathematics  0.627604  
607                        Brain and Cognitive Sciences  0.630080  
715         Electrical Engineering and Computer Science  0.671841  
173                          Sloan School of Management  0.678863  
2134                       Brain and Cognitive Sciences  0.683056  
1697                                        Mathematics  0.690946  
1519  Electrical Engineering and Computer Science, H...  0.704394  
154                        Brain and Cognitive Sciences  0.727064  
259                                         Mathematics  0.731739

BoW and TF-IDF with cosine similarity produce good results. Most of the courses recommended are similar to original course ("Machine Learning"), featuring keywords such as "Statistics", "Artificial Intelligence", "Data Analysis", and "Mathematics". 

Both methods produce almost identical results, with the main difference being the order of similarity. However, "Algorithmic Aspects of Machine Learning" is present in TF-IDF method but not in BoW method. Conversely, "Techniques in Artificial Intelligence (SMA 5504)" is present in BoW method but not in TF-IDF method.

## Modern Methods (Word2Vec, FastText, SentenceTransformers)

In [ ]:
print("Word2Vec (Pre-Trained on WikiNews) x WMD Method:")
display(w2v_wmd_dist.loc[:11, ['title', 'description', 'topics', 'department_name', 'wmd_dist']])

print("Word2Vec (Pre-Trained on WikiNews) x Cosine Method:")
display(w2v_cos_dist.loc[:11, ['title', 'description', 'topics', 'department_name', 'cos_dist']])

print("FastText (Trained on This Corpus) x WMD Method:")
display(ft_wmd_dist.loc[:11, ['title', 'description', 'topics', 'department_name', 'wmd_dist']])

print("Sentence Transformers (all-MiniLM-L6-v2) x Cosine Method:")
display(st_cos_dist.loc[:11, ['title', 'description', 'topics', 'department_name', 'cos_dist']])

Word2Vec (Pre-Trained on WikiNews) x WMD Method:


title  \
1797                                   Machine Learning   
1780   Techniques in Artificial Intelligence (SMA 5504)   
1218  Topics in Statistics: Statistical Learning Theory   
154        Statistical Learning Theory and Applications   
2134  Networks for Learning: Regression and Classifi...   
715                    Introduction to Machine Learning   
1697  Matrix Methods in Data Analysis, Signal Proces...   
607        Statistical Learning Theory and Applications   
2286                           Algorithms for Inference   
2136                           Medical Decision Support   
1519                    Machine Learning for Healthcare   

                                            description  \
1797  6.867 is an introductory course on machine lea...   
1780  6.825 is a graduate-level introduction to arti...   
1218  The main goal of this course is to study the g...   
154   This course is for upper-level graduate studen...   
2134  The course focuses on the problem of supervise...   
715   This course introduces principles, algorithms,...   
1697  Linear algebra concepts are key for understand...   
607   Focuses on the problem of supervised learning ...   
2286  This is a graduate-level introduction to the p...   
2136  This course presents the main concepts of deci...   
1519  This course introduces students to machine lea...   

                                                 topics  \
1797  Engineering, Science, Computer Science, Artifi...   
1780  Engineering, Science, Computer Science, Artifi...   
1218  Engineering, Science, Computer Science, Artifi...   
154   Science, Mathematics, Probability and Statisti...   
2134  Science, Mathematics, Probability and Statisti...   
715   Engineering, Computer Science, Artificial Inte...   
1697  Engineering, Mathematics, Electrical Engineeri...   
607   Engineering, Science, Computer Science, Artifi...   
2286  Engineering, Computer Science, Mathematics, Pr...   
2136  Engineering, Computer Science, Artificial Inte...   
1519  Engineering, Computer Science, Artificial Inte...   

                                        department_name  wmd_dist  
1797        Electrical Engineering and Computer Science  0.000000  
1780        Electrical Engineering and Computer Science  0.693753  
1218                                        Mathematics  0.702761  
154                        Brain and Cognitive Sciences  0.721380  
2134                       Brain and Cognitive Sciences  0.726982  
715         Electrical Engineering and Computer Science  0.729386  
1697                                        Mathematics  0.733940  
607                        Brain and Cognitive Sciences  0.734341  
2286        Electrical Engineering and Computer Science  0.734567  
2136  Health Sciences and Technology, Electrical Eng...  0.740701  
1519  Electrical Engineering and Computer Science, H...  0.751948

Word2Vec (Pre-Trained on WikiNews) x Cosine Method:


title  \
1797                                   Machine Learning   
1631                      Introduction to Deep Learning   
607        Statistical Learning Theory and Applications   
2001                   Pattern Recognition and Analysis   
715                    Introduction to Machine Learning   
98    Probabilistic Systems Analysis and Applied Pro...   
2245        Great Ideas in Theoretical Computer Science   
1705  Introduction to Electrical Engineering and Com...   
721               Machine Learning for Inverse Graphics   
1699  Introduction to Computational Neuroscience wit...   
2540  Generative Artificial Intelligence in K–12 Edu...   

                                            description  \
1797  6.867 is an introductory course on machine lea...   
1631  This is MIT's introductory course on deep lear...   
607   Focuses on the problem of supervised learning ...   
2001  This class deals with the fundamentals of char...   
715   This course introduces principles, algorithms,...   
98    Welcome to 6.041/6.431, a subject on the model...   
2245  This course provides a challenging introductio...   
1705  This course provides an integrated introductio...   
721   This course covers fundamental and advanced te...   
1699  In this course, you will learn the basics of c...   
2540  The emergence of transformer architectures in ...   

                                                 topics  \
1797  Engineering, Science, Computer Science, Artifi...   
1631  Engineering, Computer Science, Artificial Inte...   
607   Engineering, Science, Computer Science, Artifi...   
2001  Engineering, Science, Mathematics, Probability...   
715   Engineering, Computer Science, Artificial Inte...   
98    Engineering, Mathematics, Systems Engineering,...   
2245  Engineering, Computer Science, Mathematics, Ap...   
1705  Engineering, Computer Science, Artificial Inte...   
721   Engineering, Computer Science, Artificial Inte...   
1699  Engineering, Science, Computer Science, Health...   
2540  Engineering, Computer Science, Artificial Inte...   

                                        department_name      cos_dist  
1797        Electrical Engineering and Computer Science  5.960464e-08  
1631        Electrical Engineering and Computer Science  1.353943e-02  
607                        Brain and Cognitive Sciences  1.458144e-02  
2001  Media Arts and Sciences, Civil and Environment...  1.546454e-02  
715         Electrical Engineering and Computer Science  1.668727e-02  
98          Electrical Engineering and Computer Science  1.685762e-02  
2245        Electrical Engineering and Computer Science  1.773494e-02  
1705        Electrical Engineering and Computer Science  1.817471e-02  
721         Electrical Engineering and Computer Science  1.864868e-02  
1699                       Brain and Cognitive Sciences  1.886159e-02  
2540  Electrical Engineering and Computer Science, M...  1.913077e-02

FastText (Trained on This Corpus) x WMD Method:


title  \
1797                                   Machine Learning   
1780   Techniques in Artificial Intelligence (SMA 5504)   
1218  Topics in Statistics: Statistical Learning Theory   
607        Statistical Learning Theory and Applications   
715                    Introduction to Machine Learning   
1519                    Machine Learning for Healthcare   
154        Statistical Learning Theory and Applications   
259                     Mathematics of Machine Learning   
2134  Networks for Learning: Regression and Classifi...   
1697  Matrix Methods in Data Analysis, Signal Proces...   
173         Prediction: Machine Learning and Statistics   

                                            description  \
1797  6.867 is an introductory course on machine lea...   
1780  6.825 is a graduate-level introduction to arti...   
1218  The main goal of this course is to study the g...   
607   Focuses on the problem of supervised learning ...   
715   This course introduces principles, algorithms,...   
1519  This course introduces students to machine lea...   
154   This course is for upper-level graduate studen...   
259   Broadly speaking, Machine Learning refers to t...   
2134  The course focuses on the problem of supervise...   
1697  Linear algebra concepts are key for understand...   
173   Prediction is at the heart of almost every sci...   

                                                 topics  \
1797  Engineering, Science, Computer Science, Artifi...   
1780  Engineering, Science, Computer Science, Artifi...   
1218  Engineering, Science, Computer Science, Artifi...   
607   Engineering, Science, Computer Science, Artifi...   
715   Engineering, Computer Science, Artificial Inte...   
1519  Engineering, Computer Science, Artificial Inte...   
154   Science, Mathematics, Probability and Statisti...   
259   Engineering, Computer Science, Artificial Inte...   
2134  Science, Mathematics, Probability and Statisti...   
1697  Engineering, Mathematics, Electrical Engineeri...   
173   Engineering, Computer Science, Artificial Inte...   

                                        department_name  wmd_dist  
1797        Electrical Engineering and Computer Science  0.000000  
1780        Electrical Engineering and Computer Science  0.820359  
1218                                        Mathematics  0.820833  
607                        Brain and Cognitive Sciences  0.867170  
715         Electrical Engineering and Computer Science  0.889718  
1519  Electrical Engineering and Computer Science, H...  0.909089  
154                        Brain and Cognitive Sciences  0.913724  
259                                         Mathematics  0.916447  
2134                       Brain and Cognitive Sciences  0.920859  
1697                                        Mathematics  0.922326  
173                          Sloan School of Management  0.936024

Sentence Transformers (all-MiniLM-L6-v2) x Cosine Method:


title  \
1797                                   Machine Learning   
1218  Topics in Statistics: Statistical Learning Theory   
1780   Techniques in Artificial Intelligence (SMA 5504)   
2001                   Pattern Recognition and Analysis   
715                    Introduction to Machine Learning   
259                     Mathematics of Machine Learning   
2134  Networks for Learning: Regression and Classifi...   
173         Prediction: Machine Learning and Statistics   
154        Statistical Learning Theory and Applications   
1697  Matrix Methods in Data Analysis, Signal Proces...   
1564                            Artificial Intelligence   

                                            description  \
1797  6.867 is an introductory course on machine lea...   
1218  The main goal of this course is to study the g...   
1780  6.825 is a graduate-level introduction to arti...   
2001  This class deals with the fundamentals of char...   
715   This course introduces principles, algorithms,...   
259   Broadly speaking, Machine Learning refers to t...   
2134  The course focuses on the problem of supervise...   
173   Prediction is at the heart of almost every sci...   
154   This course is for upper-level graduate studen...   
1697  Linear algebra concepts are key for understand...   
1564  This course introduces representations, techni...   

                                                 topics  \
1797  Engineering, Science, Computer Science, Artifi...   
1218  Engineering, Science, Computer Science, Artifi...   
1780  Engineering, Science, Computer Science, Artifi...   
2001  Engineering, Science, Mathematics, Probability...   
715   Engineering, Computer Science, Artificial Inte...   
259   Engineering, Computer Science, Artificial Inte...   
2134  Science, Mathematics, Probability and Statisti...   
173   Engineering, Computer Science, Artificial Inte...   
154   Science, Mathematics, Probability and Statisti...   
1697  Engineering, Mathematics, Electrical Engineeri...   
1564  Engineering, Computer Science, Artificial Inte...   

                                        department_name      cos_dist  
1797        Electrical Engineering and Computer Science  1.192093e-07  
1218                                        Mathematics  1.994100e-01  
1780        Electrical Engineering and Computer Science  2.439162e-01  
2001  Media Arts and Sciences, Civil and Environment...  2.652936e-01  
715         Electrical Engineering and Computer Science  2.720942e-01  
259                                         Mathematics  2.878628e-01  
2134                       Brain and Cognitive Sciences  2.889742e-01  
173                          Sloan School of Management  2.956704e-01  
154                        Brain and Cognitive Sciences  3.100022e-01  
1697                                        Mathematics  3.104867e-01  
1564        Electrical Engineering and Computer Science  3.333538e-01

All modern methods produce good results, recommending items similar to original course ("Machine Laerning"). Resulting items with keywords such as "Statistics", "Artificial Intelligence", "Data Analysis", and "Mathematics". 

However, there are courses recommended that are not present in traditional methods, such as:
* "Pattern Recognition and Analysis" from ST x Cosine
* "Probabilistic Systems Analysis and Applied Probability" from W2V x Cosine
* "Algorithms for Inference" from W2V X WMD

This shows that modern methods are able to capture more semantically rich context and do not solely rely on keyword frequency, unlike traditional methods.

# Conclusion

For this problem, I believe traditional methods are preferable due to their good results and fast performance. Since the task involves only text similarity based on word occurrences, and the course descriptions are descriptive (without requiring sentiment analysis), traditional methods are sufficient.

The FastText method, even when trained on a relatively small corpus, also performs well and is comparable to pre-trained models like W2V and ST.